crowbar_framework/app/models/crowbar_service.rb

```diff
@@ -32,6 +32,86 @@ def role_constraints
     end
   end
 
+  # This is relevant to upgrade process only.
+  #
+  # Commit current proposal of crowbar barclamp and check if the commit doesn't
+  # end with some errors.
+  #
+  # Unfortunatelly we need to explicitely look at crowbar-status of the proposal
+  # because apply_role from this model ignores errors from superclass's apply_role.
+  def commit_and_check_proposal
+    proposal_commit("default", false, false)
+
+    proposal = ProposalObject.find_proposal("crowbar", "default")
+    if proposal["deployment"]["crowbar"]["crowbar-status"] == "failed"
+      raise proposal["deployment"]["crowbar"]["crowbar-failed"]
+    end
+  end
+
+  def shutdown_services_at_non_db_nodes
+    db_nodes = []
+    NodeObject.find("state:crowbar_upgrade").each do |node|
+
+      # Find the nodes with Database role
+      # (we can't look at node.roles, because that list has been reduced to just crowbar-upgrade)
+      node.crowbar["run_list_map"].each do |role, v|
+        db_nodes.push node.name if role == "database-server"
+      end
+
+      # mark the position in the upgrade process
+      node["crowbar_wall"]["crowbar_openstack_shutdown"] = true
+      node.save
+    end
+
+    # Adapt current proposal, so only non-db nodes have crowbar-upgrade role
+    # (crowbar-upgrade role is currently assigned to all nodes)
+    proposal = ProposalObject.find_proposal("crowbar", "default")
+    proposal["deployment"]["crowbar"]["elements"]["crowbar-upgrade"] -= db_nodes
+    proposal.save
+
+    return true if proposal["deployment"]["crowbar"]["elements"]["crowbar-upgrade"].empty?
+
+    # Commit proposal so the shutdown actions from crowbar-upgrade get executed for non-db nodes
+    commit_and_check_proposal
+  end
+
+  def dump_openstack_database
+    # Find the nodes with Database role
+    # (we can't look at node.roles, because that list has been reduced to just crowbar-upgrade item)
+    db_nodes = NodeObject.find("state:crowbar_upgrade AND run_list_map:database-server")
+
+    raise "There does not seem to exist any node running the database." if db_nodes.empty?
```

Comment:
> My intention was to have a check to make sure here, that all nodes returned here have the 
> right thing in ["crowbar_wall"]["crowbar_openstack_shutdown"]. Mainly to prevent 
> dump_openstack_database from doing anything before
> shutdown_services_at_non_db_nodes was called.

But again, this is checked in the recipe: https://github.com/crowbar/crowbar-core/blob/master/chef/cookbooks/crowbar/recipes/crowbar-db-dump.rb#L30 (and the else branch)